src/async.rs

```diff
@@ -0,0 +1,228 @@
+//! Provides asynchronous retry functionality via `futures`.
+//!
+//! # Examples
+//!
+//! ```rust
+//! # extern crate futures;
+//! # extern crate tokio_timer;
+//! # extern crate retry;
+//! # use std::default::Default;
+//! # use futures::future::Future;
+//! use tokio_timer::Timer;
+//! use retry::delay::{Exponential, jitter};
+//! use retry::async::retry;
+//!
+//! pub fn main() {
+//!     let timer = Timer::default();
+//!     let strategy = Exponential::from_millis(10)
+//!         .map(jitter).take(3);
+//!
+//!     let mut collection = vec![1, 2, 3].into_iter();
+//!
+//!     let future = retry(timer, strategy, || match collection.next() {
+//!         Some(n) if n == 3 => Ok("n is 3!"),
+//!         Some(_) => Err("n must be 3!"),
+//!         None => Err("n was never 3!"),
+//!     });
+//!
+//!     let result = future.wait();
+//!
+//!     assert!(result.is_ok());
+//! }
+//! ```
+
+use either::Either;
+use futures::{Async, IntoFuture, Future, Poll};
+use futures::future::{Flatten, FutureResult};
+use std::error::Error;
+use std::io;
+use std::fmt;
+use std::time::Duration;
+#[cfg(feature = "async_tokio_timer")]
+use tokio_timer;
+#[cfg(feature = "async_tokio_core")]
+use tokio_core::reactor;
+
+use super::{Error as RetryError};
+
+/// Trait that represents being able to produce a future that resolves
+/// after a given duration.
```

Comment:
I was actually trying to follow the [guidelines on naming](https://aturon.github.io/style/naming.html) you posted last time, and preferred a verb over a noun here.

`Timer` is difficult imho because this is already the name of one of the structs that implements this trait (`tokio_timer::Timer`), which might be confusing. `FutureDelay` seems to imply it's a delay in some way; from the looks of it we have already settled on a "delay" being an `Iterator<Duration>`, and I would hate for it to mean two different things.